### PR TITLE
Fix sending redundant RequestN frame

### LINKED ISSUE #97: Bug with flow.buffer(2).take(2)
From https://github.com/yschimke/rsocket-cli/pull/1

Total request is 4

```
Send:
RequestStream frame -> Stream Id: 1 Length: 13
Flags: 0b100000000 (M1F0C0N0)
Initial request: 2
Metadata: Empty
Data: Empty


Receive:
Payload frame -> Stream Id: 1 Length: 19
Flags: 0b000100000 (M0F0C0N1)
Data(length=13):
         +-------------------------------------------------+
         |  0  1  2  3  4  5  6  7  8  9  a  b  c  d  e  f |
+--------+-------------------------------------------------+----------------+
|00000000| 31 35 39 39 38 35 30 30 39 31 34 31 34          |1599850091414   |
+--------+-------------------------------------------------+----------------+

1599850091414

Receive:
Payload frame -> Stream Id: 1 Length: 19
Flags: 0b000100000 (M0F0C0N1)
Data(length=13):
         +-------------------------------------------------+
         |  0  1  2  3  4  5  6  7  8  9  a  b  c  d  e  f |
+--------+-------------------------------------------------+----------------+
|00000000| 31 35 39 39 38 35 30 30 39 31 34 36 34          |1599850091464   |
+--------+-------------------------------------------------+----------------+


Send:
RequestN frame -> Stream Id: 1 Length: 10
Flags: 0b000000000 ()
RequestN: 2

1599850091464

Send:
Cancel frame -> Stream Id: 1 Length: 6
Flags: 0b000000000 ()
```

### PR DESCRIPTION
**Pull request by whyoleg**
fixes #97

### CODE REVIEW

File: rsocket-core/build.gradle.kts [hunk 1]
```diff
@@ -36,6 +36,7 @@ kotlin {
         }
         val commonTest by getting {
             dependencies {
+                implementation("app.cash.turbine:turbine:0.2.1")
```
Comment (yschimke) [inline rsocket-core/build.gradle.kts:39]: nice!

Comment (whyoleg) [reply rsocket-core/build.gradle.kts:39]: yeah, first try of that library, looks really good!

File: rsocket-core/src/commonMain/kotlin/io/rsocket/kotlin/internal/flow/RequestChannelRequesterFlow.kt [hunk 2]
```diff
@@ -35,8 +35,7 @@ internal class RequestChannelRequesterFlow(
     override fun create(context: CoroutineContext, capacity: Int): RequestChannelRequesterFlow =
         RequestChannelRequesterFlow(payloads, requester, state, context, capacity)
 
-    @OptIn(ExperimentalCoroutinesApi::class)
-    override suspend fun collectTo(scope: ProducerScope<Payload>): Unit = with(state) {
+    override suspend fun collectImpl(collectContext: CoroutineContext, collector: FlowCollector<Payload>): Unit = with(state) {
```
Comment (yschimke) [inline rsocket-core/src/commonMain/kotlin/io/rsocket/kotlin/internal/flow/RequestChannelRequesterFlow.kt:38]: Is this the new guidance for accessing the coroutineContext? There have been various appraoches in past few years.

https://kotlinlang.org/docs/reference/coroutines/basics.html

File: rsocket-core/src/commonMain/kotlin/io/rsocket/kotlin/internal/flow/StreamFlow.kt [hunk 3]
```diff
@@ -40,18 +41,33 @@ internal abstract class StreamFlow(
...
     ): Unit = with(state) {
-        val collector = SendingCollector(scope.channel)
         consumeReceiverFor(streamId) {
             var consumed = 0
             //TODO fragmentation
             for (frame in receiver) {
                 if (frame.complete) return //TODO check next flag
-                collector.emit(frame.payload)
+                //emit in collectContext to prevent `Flow invariant is violated`
+                withContext(collectContext) {
```
Comment (yschimke) [inline rsocket-core/src/commonMain/kotlin/io/rsocket/kotlin/internal/flow/StreamFlow.kt:68]: Is this the core of the fix?

**Review by yschimke — APPROVED** [review]
I don't understand the fix, but since it adds additional tests without breaking existing ⛴

**Comment by whyoleg** [comment]
The idea of the fix:
`StreamFlow` inherits `ChannelFlow` to support flow fusion (so to support `buffer(n)` operator). Before, it was using channels under the hood. Because of it, additional requestN can be sent, because of asynchronous cancellation. Now it don't use channels, but manually switch context (needed to support `flowOn` operator). But because emiting to FlowCollector should be from the same dispatcher, we need to switch context back while emiting.

**Pull request merged**